File: apps/api/src/chatbot_api/api/v1/endpoints/conversations.py
from datetime import date

from fastapi import APIRouter, Depends, Query, status
from sqlalchemy.ext.asyncio import AsyncSession

from chatbot_api.api.dependencies import get_current_admin
from chatbot_api.core.db import get_session
from chatbot_api.models import Admin
from chatbot_api.models.enums import ConversationStatus
from chatbot_api.schemas.conversation import (
    ContactUpdate,
    ConversationDetail,
    ConversationHistory,
    ConversationListItem,
    SendMessageRequest,
    SendMessageResponse,
    StarUpdate,
)
from chatbot_api.schemas.internal_note import NoteCreate, NoteRead, NoteUpdate
from chatbot_api.schemas.message import MessageRead
from chatbot_api.schemas.pagination import Page, PageParams
from chatbot_api.schemas.tag import TagAssign
from chatbot_api.services import conversation_service, note_service, tag_service

router = APIRouter(prefix="/conversations", tags=["conversations"])


@router.get("", response_model=Page[ConversationListItem])
async def list_conversations(
    status_filter: ConversationStatus | None = Query(None, alias="status"),
    from_date: date | None = None,
    to_date: date | None = None,
    phone: str | None = None,
    page: int = Query(1, ge=1),
    size: int = Query(20, ge=1, le=100),
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> Page[ConversationListItem]:
    return await conversation_service.list_paginated(
        db,
        status=status_filter,
        from_date=from_date,
        to_date=to_date,
        phone=phone,
        pagination=PageParams(page=page, size=size),
    )


@router.get("/{conversation_id}", response_model=ConversationDetail)
async def get_conversation(
    conversation_id: int,
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> ConversationDetail:
    return await conversation_service.get_detail(db, conversation_id)


@router.get("/{conversation_id}/messages", response_model=Page[MessageRead])
async def list_messages(
    conversation_id: int,
    page: int = Query(1, ge=1),
    size: int = Query(50, ge=1, le=200),
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> Page[MessageRead]:
    return await conversation_service.list_messages_paginated(
        db, conversation_id, pagination=PageParams(page=page, size=size)
    )


@router.post("/{conversation_id}/takeover")
async def takeover(
    conversation_id: int,
    admin: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> dict[str, object]:
    return await conversation_service.takeover(
        db, conversation_id=conversation_id, admin_id=admin.id
    )


@router.post("/{conversation_id}/release")
async def release(
    conversation_id: int,
    admin: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> dict[str, object]:
    return await conversation_service.release(
        db, conversation_id=conversation_id, admin_id=admin.id
    )


@router.post("/{conversation_id}/close")
async def close(
    conversation_id: int,
    admin: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> dict[str, object]:
    return await conversation_service.close(
        db, conversation_id=conversation_id, admin_id=admin.id
    )


@router.post("/{conversation_id}/reopen")
async def reopen(
    conversation_id: int,
    admin: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> dict[str, object]:
    return await conversation_service.reopen(
        db, conversation_id=conversation_id, admin_id=admin.id
    )


@router.delete("/{conversation_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_conversation(
    conversation_id: int,
    admin: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> None:
    await conversation_service.delete_conversation(
        db, conversation_id=conversation_id, admin_id=admin.id
    )


@router.post(
    "/{conversation_id}/messages",
    response_model=SendMessageResponse,
    status_code=status.HTTP_201_CREATED,
)
async def send_message(
    conversation_id: int,
    payload: SendMessageRequest,
    admin: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> SendMessageResponse:
    return await conversation_service.send_admin_message(
        db,
        conversation_id=conversation_id,
        body=payload.body,
        admin_id=admin.id,
        in_reply_to_id=payload.in_reply_to_id,
    )


# ── Ficha de contacto: correo, destacar, historial ────────────────────


@router.get("/{conversation_id}/history", response_model=ConversationHistory)
async def get_history(
    conversation_id: int,
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> ConversationHistory:
    return await conversation_service.get_history(db, conversation_id)


@router.patch("/{conversation_id}/contact", response_model=ConversationDetail)
async def update_contact(
    conversation_id: int,
    payload: ContactUpdate,
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> ConversationDetail:
    return await conversation_service.update_contact(
        db, conversation_id=conversation_id, email=payload.email
    )


@router.put("/{conversation_id}/star", response_model=ConversationDetail)
async def set_star(
    conversation_id: int,
    payload: StarUpdate,
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> ConversationDetail:
    return await conversation_service.set_starred(
        db, conversation_id=conversation_id, starred=payload.starred
    )


# ── Etiquetas de la conversación ──────────────────────────────────────


@router.post("/{conversation_id}/tags", response_model=ConversationDetail)
async def assign_tag(
    conversation_id: int,
    payload: TagAssign,
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> ConversationDetail:
    return await tag_service.assign_tag(
        db, conversation_id=conversation_id, tag_id=payload.tag_id
    )


@router.delete("/{conversation_id}/tags/{tag_id}", response_model=ConversationDetail)
async def unassign_tag(
    conversation_id: int,
    tag_id: int,
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> ConversationDetail:
    return await tag_service.unassign_tag(
        db, conversation_id=conversation_id, tag_id=tag_id
    )


# ── Notas internas ────────────────────────────────────────────────────


@router.get("/{conversation_id}/notes", response_model=list[NoteRead])
async def list_notes(
    conversation_id: int,
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> list[NoteRead]:
    return await note_service.list_notes(db, conversation_id)


@router.post(
    "/{conversation_id}/notes",
    response_model=NoteRead,
    status_code=status.HTTP_201_CREATED,
)
async def create_note(
    conversation_id: int,
    payload: NoteCreate,
    admin: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> NoteRead:
    return await note_service.create_note(
        db, conversation_id=conversation_id, author=admin, body=payload.body
    )


@router.patch("/{conversation_id}/notes/{note_id}", response_model=NoteRead)
async def update_note(
    conversation_id: int,
    note_id: int,
    payload: NoteUpdate,
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> NoteRead:
    return await note_service.update_note(
        db, conversation_id=conversation_id, note_id=note_id, body=payload.body
    )


@router.delete(
    "/{conversation_id}/notes/{note_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_note(
    conversation_id: int,
    note_id: int,
    _: Admin = Depends(get_current_admin),
    db: AsyncSession = Depends(get_session),
) -> None:
    await note_service.delete_note(
        db, conversation_id=conversation_id, note_id=note_id
    )

File: apps/web/src/lib/api/conversations.ts
import { apiFetch } from "./client";

export type ConversationStatus = "abierta" | "cerrada" | "takeover";
export type MessageRole = "student" | "bot" | "admin";

export interface StudentRef {
  phone_e164: string;
  display_name: string | null;
}

export interface Tag {
  id: number;
  name: string;
  color: string;
}

export interface ConversationListItem {
  id: number;
  student_phone: string;
  display_name: string | null;
  status: ConversationStatus;
  opened_at: string;
  closed_at: string | null;
  message_count: number;
  last_message_preview: string | null;
  starred: boolean;
}

export interface QuotedSnapshot {
  id: number;
  role: MessageRole;
  content: string;
  created_at: string | null;
}

export interface MessageRead {
  id: number;
  conversation_id: number;
  role: MessageRole;
  content: string;
  intent_id: number | null;
  retrieved_chunks: unknown[];
  input_tokens: number | null;
  output_tokens: number | null;
  model_used: string | null;
  latency_ms: number | null;
  meta_message_id: string | null;
  created_at: string;
  in_reply_to_id: number | null;
  quoted: QuotedSnapshot | null;
  delivery_status: DeliveryStatus | null;
}

export type DeliveryStatus = "sent" | "delivered" | "read" | "failed";

export interface StudentProfile {
  phone_e164: string;
  full_name: string;
  career: string | null;
  cycle: number | null;
  campus: string | null;
  modality: string | null;
  academic_status: string | null;
  failed_courses: string | null;
  enrollment_turn: string | null;
  english_level: number | null;
  elective_credits: number | null;
  internship_credits: number | null;
}

export interface ConversationDetail {
  id: number;
  student_phone: string;
  display_name: string | null;
  email: string | null;
  status: ConversationStatus;
  opened_at: string;
  closed_at: string | null;
  takeover_admin: number | null;
  starred: boolean;
  student_profile: StudentProfile | null;
  tags: Tag[];
  messages: MessageRead[];
}

export interface InternalNote {
  id: number;
  body: string;
  author_admin_id: number | null;
  author_name: string | null;
  created_at: string;
  updated_at: string;
}

export interface ConversationHistory {
  total_conversations: number;
  total_messages: number;
  first_contact: string | null;
}

export interface Page<T> {
  items: T[];
  total: number;
  page: number;
  size: number;
  pages: number;
}

export interface FetchConversationsParams {
  status?: ConversationStatus;
  phone?: string;
  page?: number;
  size?: number;
}

export async function fetchConversations(
  params: FetchConversationsParams = {},
): Promise<Page<ConversationListItem>> {
  return apiFetch<Page<ConversationListItem>>("/api/v1/conversations", {
    searchParams: {
      status: params.status,
      phone: params.phone,
      page: params.page ?? 1,
      size: params.size ?? 20,
    },
  });
}

export async function fetchConversation(
  id: number,
): Promise<ConversationDetail> {
  return apiFetch<ConversationDetail>(`/api/v1/conversations/${id}`);
}

export async function fetchMessages(
  conversationId: number,
  params: { page?: number; size?: number } = {},
): Promise<Page<MessageRead>> {
  return apiFetch<Page<MessageRead>>(
    `/api/v1/conversations/${conversationId}/messages`,
    {
      searchParams: {
        page: params.page ?? 1,
        size: params.size ?? 100,
      },
    },
  );
}

export interface SendMessageResponse {
  message_id: number;
  meta_message_id: string | null;
  conversation_status: ConversationStatus;
}

export async function sendConversationMessage(
  conversationId: number,
  body: string,
  inReplyToId?: number | null,
): Promise<SendMessageResponse> {
  return apiFetch<SendMessageResponse>(
    `/api/v1/conversations/${conversationId}/messages`,
    {
      method: "POST",
      body: JSON.stringify({ body, in_reply_to_id: inReplyToId ?? null }),
    },
  );
}

export async function takeoverConversation(
  conversationId: number,
): Promise<unknown> {
  return apiFetch<unknown>(`/api/v1/conversations/${conversationId}/takeover`, {
    method: "POST",
  });
}

export async function releaseConversation(
  conversationId: number,
): Promise<unknown> {
  return apiFetch<unknown>(`/api/v1/conversations/${conversationId}/release`, {
    method: "POST",
  });
}

export async function closeConversation(
  conversationId: number,
): Promise<unknown> {
  return apiFetch<unknown>(`/api/v1/conversations/${conversationId}/close`, {
    method: "POST",
  });
}

export async function reopenConversation(
  conversationId: number,
): Promise<unknown> {
  return apiFetch<unknown>(`/api/v1/conversations/${conversationId}/reopen`, {
    method: "POST",
  });
}

export async function deleteConversation(
  conversationId: number,
): Promise<void> {
  return apiFetch<void>(`/api/v1/conversations/${conversationId}`, {
    method: "DELETE",
  });
}

// ── Ficha de contacto ────────────────────────────────────────────────

export async function updateContact(
  conversationId: number,
  email: string | null,
): Promise<ConversationDetail> {
  return apiFetch<ConversationDetail>(
    `/api/v1/conversations/${conversationId}/contact`,
    { method: "PATCH", body: JSON.stringify({ email }) },
  );
}

export async function setStar(
  conversationId: number,
  starred: boolean,
): Promise<ConversationDetail> {
  return apiFetch<ConversationDetail>(
    `/api/v1/conversations/${conversationId}/star`,
    { method: "PUT", body: JSON.stringify({ starred }) },
  );
}

export async function fetchHistory(
  conversationId: number,
): Promise<ConversationHistory> {
  return apiFetch<ConversationHistory>(
    `/api/v1/conversations/${conversationId}/history`,
  );
}

// ── Notas internas ───────────────────────────────────────────────────

export async function fetchNotes(
  conversationId: number,
): Promise<InternalNote[]> {
  return apiFetch<InternalNote[]>(
    `/api/v1/conversations/${conversationId}/notes`,
  );
}

export async function createNote(
  conversationId: number,
  body: string,
): Promise<InternalNote> {
  return apiFetch<InternalNote>(
    `/api/v1/conversations/${conversationId}/notes`,
    { method: "POST", body: JSON.stringify({ body }) },
  );
}

export async function updateNote(
  conversationId: number,
  noteId: number,
  body: string,
): Promise<InternalNote> {
  return apiFetch<InternalNote>(
    `/api/v1/conversations/${conversationId}/notes/${noteId}`,
    { method: "PATCH", body: JSON.stringify({ body }) },
  );
}

export async function deleteNote(
  conversationId: number,
  noteId: number,
): Promise<void> {
  return apiFetch<void>(
    `/api/v1/conversations/${conversationId}/notes/${noteId}`,
    { method: "DELETE" },
  );
}

// ── Etiquetas de la conversación ─────────────────────────────────────

export async function assignTag(
  conversationId: number,
  tagId: number,
): Promise<ConversationDetail> {
  return apiFetch<ConversationDetail>(
    `/api/v1/conversations/${conversationId}/tags`,
    { method: "POST", body: JSON.stringify({ tag_id: tagId }) },
  );
}

export async function unassignTag(
  conversationId: number,
  tagId: number,
): Promise<ConversationDetail> {
  return apiFetch<ConversationDetail>(
    `/api/v1/conversations/${conversationId}/tags/${tagId}`,
    { method: "DELETE" },
  );
}
